Confusing error message when using #[pyclass] with a struct called Probe

### Bug Description

The following code fails to compile, with a somewhat confusing error message:

```rust
use pyo3::prelude::*;

#[pyclass]
struct Probe {}
```

The error message is:

```
error[E0782]: expected a type, found a trait
 --> src/lib.rs:4:8
  |
4 | struct Probe {}
  |        ^^^^^
  |
help: you can add the `dyn` keyword if you want a trait object
  |
4 | struct dyn Probe {}
  |        +++

For more information about this error, try `rustc --explain E0782`.
error: could not compile `pyo3-repro` (lib) due to 1 previous error
```

rust-analyzer suggests that this might be due to the `trait Probe` here:
https://github.com/PyO3/pyo3/blob/926f5cf8e427d08b0a4670d89e3b0f342c299489/src/impl_/pyclass/probes.rs#L16

I'm not sure if this can be fixed, but it would be nice to at least have a clearer error message to indicate that the struct can't be called `Probe`. 

### Steps to Reproduce

Compile the code mentioned above.

### Backtrace

_No response_

### Your operating system and version

Ubuntu 24.04

### Your Python version (`python --version`)

Python 3.12.3

### Your Rust version (`rustc --version`)

rustc 1.83.0 (90b35a623 2024-11-26)

### Your PyO3 version

0.23.3

### How did you install python? Did you use a virtualenv?

apt

### Additional Info

This was observed when updating PyO3 from version 0.21 to 0.22, so it seems to have changed in version 0.22.

--- a/newsfragments/4794.fixed.md
+++ b/newsfragments/4794.fixed.md
@@ -1,0 +1,1 @@
+fix `#[pyclass]` could not be named `Probe`

--- a/pyo3-macros-backend/src/pyclass.rs
+++ b/pyo3-macros-backend/src/pyclass.rs
@@ -2294,10 +2294,9 @@ impl<'a> PyClassImplsBuilder<'a> {
         let assertions = if attr.options.unsendable.is_some() {
             TokenStream::new()
         } else {
-            let assert = quote_spanned! { cls.span() => assert_pyclass_sync::<#cls>(); };
+            let assert = quote_spanned! { cls.span() => #pyo3_path::impl_::pyclass::assert_pyclass_sync::<#cls>(); };
             quote! {
                 const _: () = {
-                    use #pyo3_path::impl_::pyclass::*;
                     #assert
                 };
             }
@@ -2337,7 +2336,7 @@ impl<'a> PyClassImplsBuilder<'a> {
                     static DOC: #pyo3_path::sync::GILOnceCell<::std::borrow::Cow<'static, ::std::ffi::CStr>> = #pyo3_path::sync::GILOnceCell::new();
                     DOC.get_or_try_init(py, || {
                         let collector = PyClassImplCollector::<Self>::new();
-                        build_pyclass_doc(<#cls as #pyo3_path::PyTypeInfo>::NAME, #doc, collector.new_text_signature())
+                        build_pyclass_doc(<Self as #pyo3_path::PyTypeInfo>::NAME, #doc, collector.new_text_signature())
                     }).map(::std::ops::Deref::deref)
                 }
 

--- a/tests/test_compile_error.rs
+++ b/tests/test_compile_error.rs
@@ -67,4 +67,5 @@ fn test_compile_errors() {
     #[cfg(all(not(Py_LIMITED_API), Py_3_11))]
     t.compile_fail("tests/ui/invalid_base_class.rs");
     t.pass("tests/ui/ambiguous_associated_items.rs");
+    t.pass("tests/ui/pyclass_probe.rs");
 }

--- a/tests/ui/pyclass_probe.rs
+++ b/tests/ui/pyclass_probe.rs
@@ -1,0 +1,6 @@
+use pyo3::prelude::*;
+
+#[pyclass]
+pub struct Probe {}
+
+fn main() {}

--- a/tests/ui/pyclass_send.stderr
+++ b/tests/ui/pyclass_send.stderr
@@ -10,7 +10,7 @@ note: required because it appears within the type `NotSyncNotSend`
   |
 5 | struct NotSyncNotSend(*mut c_void);
   |        ^^^^^^^^^^^^^^
-note: required by a bound in `pyo3::impl_::pyclass::assertions::assert_pyclass_sync`
+note: required by a bound in `assert_pyclass_sync`
  --> src/impl_/pyclass/assertions.rs
   |
   | pub const fn assert_pyclass_sync<T>()
@@ -52,7 +52,7 @@ note: required because it appears within the type `SendNotSync`
   |
 8 | struct SendNotSync(*mut c_void);
   |        ^^^^^^^^^^^
-note: required by a bound in `pyo3::impl_::pyclass::assertions::assert_pyclass_sync`
+note: required by a bound in `assert_pyclass_sync`
  --> src/impl_/pyclass/assertions.rs
   |
   | pub const fn assert_pyclass_sync<T>()

